src/interface/contract.rs

```diff
@@ -154,21 +155,23 @@ impl StateChange for AmountChange {
         *self = match self {
             AmountChange::Dec(neg) => AmountChange::Dec(*neg + sub),
             AmountChange::Zero => AmountChange::Dec(sub),
-            AmountChange::Inc(pos) if *pos > sub => AmountChange::Inc(*pos - sub),
-            AmountChange::Inc(pos) if *pos == sub => AmountChange::Zero,
-            AmountChange::Inc(pos) if *pos < sub => AmountChange::Dec(sub - *pos),
-            AmountChange::Inc(_) => unreachable!(),
+            AmountChange::Inc(pos) => match sub.cmp(pos) {
+                Ordering::Less => AmountChange::Inc(*pos - sub),
+                Ordering::Equal => AmountChange::Zero,
+                Ordering::Greater => AmountChange::Dec(sub - *pos),
+            },
         };
     }
 
     fn merge_received(&mut self, add: Self::State) {
         *self = match self {
             AmountChange::Inc(pos) => AmountChange::Inc(*pos + add),
             AmountChange::Zero => AmountChange::Inc(add),
-            AmountChange::Dec(neg) if *neg > add => AmountChange::Dec(*neg - add),
-            AmountChange::Dec(neg) if *neg == add => AmountChange::Zero,
-            AmountChange::Dec(neg) if *neg < add => AmountChange::Inc(add - *neg),
-            AmountChange::Dec(_) => unreachable!(),
+            AmountChange::Dec(neg) => match add.cmp(neg) {
+                Ordering::Greater => AmountChange::Dec(*neg - add),
```

Comment:
This must be less, not greater, since we reverse the order `neg` and `add` are compared

```suggestion
                Ordering::Less => AmountChange::Dec(*neg - add),
```